### PR TITLE
feat(compose): widgets implementations

### PR DESCRIPTION
**Pull request by Aallam**
- [x] Loading
- [x] Filter List
- [x] Hierarchical Menu
- [x] Filter Numeric Range
- [x] Current Filters
- [x] Filter Toggle
- [x] Clear Filters
- [x] Stats
- [x] Query Rule Custom Data
- [x] SortBy
- [x] Relevant Sort
- [x] Hits

### CODE REVIEW

File: instantsearch-compose/src/main/java/com/algolia/instantsearch/compose/searchbox/SearchQuery.kt [hunk 1]
```diff
@@ -21,15 +19,7 @@ public interface SearchQuery : SearchBoxView {
...
-    public fun onValueChange(query: String, isSubmit: Boolean) {
-        val onQuery = if (isSubmit) onQuerySubmitted else onQueryChanged
-        onQuery?.invoke(query)
-    }
-
-    override fun setText(text: String?, submitQuery: Boolean) {
-        query = text ?: ""
-        if (submitQuery) onQuerySubmitted?.invoke(text)
-    }
+    public fun valueChanged(query: String, isSubmit: Boolean)
```
Comment (VladislavFitz) [inline instantsearch-compose/src/main/java/com/algolia/instantsearch/compose/searchbox/SearchQuery.kt:22]:
```suggestion
    public fun changeValue(query: String, isSubmit: Boolean)
```

File: instantsearch-compose/src/main/java/com/algolia/instantsearch/compose/searchbox/internal/SearchBoxComposeImpl.kt [hunk 2]
```diff
@@ -16,4 +16,20 @@ internal class SearchBoxComposeImpl(
...
     override var onQueryChanged: Callback<String?>? = null
     override var onQuerySubmitted: Callback<String?>? = null
+
+    /**
+     * the callback to be triggered on text update
+     *
+     * @param query received query input
+     * @param isSubmit true if the value is a submit, otherwise false
+     */
+    override fun valueChanged(query: String, isSubmit: Boolean) {
```
Comment (VladislavFitz) [inline instantsearch-compose/src/main/java/com/algolia/instantsearch/compose/searchbox/internal/SearchBoxComposeImpl.kt:26]:
```suggestion
    override fun changeValue(query: String, isSubmit: Boolean) {
```

File: instantsearch-compose/src/main/java/com/algolia/instantsearch/compose/sortby/SortByState.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * Index of selected option or `null` if none is selected.
+     */
+    public val selectedOption: Int?
+
+    /**
+     * Callback on sort by option selected.
+     */
+    public fun optionSelected(index: Int?)
```
Comment (VladislavFitz) [inline instantsearch-compose/src/main/java/com/algolia/instantsearch/compose/sortby/SortByState.kt:24]:
```suggestion
    public fun selectOption(index: Int?)
```

File: instantsearch-compose/src/main/java/com/algolia/instantsearch/compose/sortby/internal/SortByStateImpl.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    override fun setMap(map: Map<Int, String>) {
+        this.options = map
+    }
+
+    override fun setSelected(selected: Int?) {
+        this.selectedOption = selected
+    }
+
+    override fun optionSelected(index: Int?) {
```
Comment (VladislavFitz) [inline instantsearch-compose/src/main/java/com/algolia/instantsearch/compose/sortby/internal/SortByStateImpl.kt:32]:
```suggestion
    override fun selectOption(index: Int?) {
```

**Review by VladislavFitz — COMMENTED** [review]
Great job! I added a few comments regarding the naming, otherwise looks good. 
Nice finding to change the `Compose` suffix to the `State`, as if I understood it right, developers will use it within `mutableStateOf(instantSearchViewState)` or `instantSearchViewState.observeAsState()` constructions.

**Comment by Aallam** [comment]
> Great job! I added a few comments regarding the naming, otherwise looks good.
> Nice finding to change the `Compose` suffix to the `State`, as if I understood it right, developers will use it within `mutableStateOf(instantSearchViewState)` or `instantSearchViewState.observeAsState()` constructions.

Yes, the user can now use the states without caring how it's implemented, and every change to them will trigger the recomposition of the view.

**Pull request merged**